File: felicia/examples/learn/message_communication/ros_msg/python/ros_msg_publishing_node.py
import felicia_py as fel
from felicia.core.protobuf.channel_pb2 import ChannelDef
from felicia.core.protobuf.master_data_pb2 import TopicInfo

from felicia.examples.learn.message_communication.ros_msg._SimpleMessage import SimpleMessage


class RosMsgPublishingNode(fel.NodeLifecycle):
    def __init__(self, node_create_flag):
        super().__init__()
        self.topic = node_create_flag.topic_flag.value
        self.channel_def_type = ChannelDef.Type.Value(node_create_flag.channel_type_flag.value)
        self.publisher = fel.communication.Publisher()
        self.publisher.set_message_impl_type(TopicInfo.ROS)
        self.message_id = 0
        self.timestamper = fel.Timestamper()

    def on_did_create(self, node_info):
        self.node_info = node_info
        self.request_publish()

    def on_request_publish(self, status):
        fel.log_if(fel.ERROR, not status.ok(), status.error_message())
        self.repeating_publish()

    def request_publish(self):
        settings = fel.communication.Settings()
        settings.buffer_size = fel.Bytes.from_bytes(512)

        self.publisher.request_publish(self.node_info, self.topic, self.channel_def_type,
                                       SimpleMessage._type,
                                       settings, self.on_request_publish)

    def repeating_publish(self):
        self.publisher.publish(self.generate_message())

        if not self.publisher.is_unregistered():
            fel.MasterProxy.post_delayed_task(
                self.repeating_publish, fel.TimeDelta.from_seconds(1))

    def generate_message(self):
        message = SimpleMessage()
        timestamp = self.timestamper.timestamp()
        message.id = self.message_id
        message.timestamp = timestamp.in_microseconds()
        message.content = "hello world"
        self.message_id += 1
        return message

File: felicia/examples/learn/message_communication/ros_msg/cc/ros_msg_publishing_node.cc
#include "felicia/examples/learn/message_communication/ros_msg/cc/ros_msg_publishing_node.h"

#include "felicia/core/master/master_proxy.h"

namespace felicia {

RosMsgPublishingNode::RosMsgPublishingNode(
    const NodeCreateFlag& node_create_flag)
    : node_create_flag_(node_create_flag),
      topic_(node_create_flag_.topic_flag()->value()) {}

void RosMsgPublishingNode::OnDidCreate(const NodeInfo& node_info) {
  node_info_ = node_info;
  RequestPublish();
}

void RosMsgPublishingNode::OnRequestPublish(const Status& s) {
  LOG_IF(ERROR, !s.ok()) << s;
  RepeatingPublish();
}

void RosMsgPublishingNode::RequestPublish() {
  communication::Settings settings;
  settings.buffer_size = Bytes::FromBytes(512);

  ChannelDef::Type channel_type;
  ChannelDef::Type_Parse(node_create_flag_.channel_type_flag()->value(),
                         &channel_type);

  publisher_.RequestPublish(
      node_info_, topic_, channel_type, settings,
      base::BindOnce(&RosMsgPublishingNode::OnRequestPublish,
                     base::Unretained(this)));
}

void RosMsgPublishingNode::RepeatingPublish() {
  publisher_.Publish(GenerateMessage());

  if (!publisher_.IsUnregistered()) {
    MasterProxy& master_proxy = MasterProxy::GetInstance();
    master_proxy.PostDelayedTask(
        FROM_HERE,
        base::BindOnce(&RosMsgPublishingNode::RepeatingPublish,
                       base::Unretained(this)),
        base::TimeDelta::FromSeconds(1));
  }
}

SimpleMessage RosMsgPublishingNode::GenerateMessage() {
  static int id = 0;
  base::TimeDelta timestamp = timestamper_.timestamp();
  SimpleMessage message;
  message.id = id++;
  message.timestamp = timestamp.InMicroseconds();
  message.content = "hello world";
  return message;
}

}  // namespace felicia